### DISCUSSION TITLE
How to Configure Scalar with Minimal API to Support Cookie Authentication in .NET 9

### DISCUSSION THREAD

**Discussion by neozhu** [scalar/scalar#3920] [Q&A]
Hello everyone,

I am currently working on a project using .NET 9 with a Minimal API approach, and I am generating API documentation using Scalar (OpenAPI). The authentication method for my endpoints is based on cookies, and I need Scalar to automatically support and retain the authentication cookie obtained from a login request for all subsequent API requests made through the Scalar UI.

Here’s what I've attempted so far:

1. **Configuration for Cookie-based Authentication in OpenAPI Security Scheme**:
   
```c#
 public static OpenApiOptions UseCookieAuthentication(this OpenApiOptions options)
{
    var scheme = new OpenApiSecurityScheme
    {
        Name = IdentityConstants.ApplicationScheme,
        Type = SecuritySchemeType.ApiKey,
        In = ParameterLocation.Cookie,
        Description = "Use this cookie to authenticate the user.",
        Reference = new OpenApiReference
        {
            Type = ReferenceType.SecurityScheme,
            Id = IdentityConstants.ApplicationScheme
        }
    };
    options.AddDocumentTransformer((document, context, ct) =>
    {
        document.Components ??= new();
        document.Components.SecuritySchemes.Add(IdentityConstants.ApplicationScheme, scheme);
        return Task.CompletedTask;
    });
    options.AddOperationTransformer((operation, context, ct) =>
    {
        if (context.Description.ActionDescriptor.EndpointMetadata.OfType<IAuthorizeData>().Any())
        {
            operation.Security = new List<OpenApiSecurityRequirement>
        {
            new OpenApiSecurityRequirement
            {
                [scheme] = new List<string>()
            }
        };
        }
        return Task.CompletedTask;
    });
    return options;
}
```

2. **Configuring AddOpenApi service scalar UI to Support Cookies**:
  
```csharp
 builder.Services.AddOpenApi(options =>
{
    options.OpenApiVersion = OpenApiSpecVersion.OpenApi3_0;
    options.UseCookieAuthentication();
    options.UseExamples();
});

app.MapOpenApi();
app.MapScalarApiReference();
```

**The Issue**:
Despite these configurations, I am still facing issues when attempting to make Scalar automatically use the authentication cookie from the login response for all subsequent API requests. Here are my specific questions:

1. **Is there a recommended way to configure Scalar with .NET 9 Minimal APIs so that it automatically retains and uses the cookie for authenticated endpoints?**
2. **Do I need to manually add the Cookie in every subsequent request made via Scalar, or is there an automatic mechanism I can leverage for this scenario?**
3. **Are there any special considerations for Minimal APIs in .NET 9 that I need to take into account for OpenAPI authentication using cookies?**

login success return and set cookie in HTTP request
![image](https://github.com/user-attachments/assets/8524b4a5-e086-44b3-998c-c9fffbf0d4ce)
next how to set cookie to HTTP request header
![image](https://github.com/user-attachments/assets/cf51ff59-42cd-48cb-a254-89e5571010d8)

I'd appreciate it if anyone could point me in the right direction, share examples, or let me know if there is a feature in Scalar that can help solve this issue effectively.

Thanks in advance for your assistance!

**Comment by xC0dex** [comment]
Hey @neozhu,

Unfortunately, I don't have much experience with cookie-based authentication, so I cannot guide you in the right direction. **But** I spot an issue which may relate to your problem: https://github.com/scalar/scalar/issues/3701
